packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -100,17 +100,123 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// Represents a range of content for a given [SelectedContent].
+///
+/// This class stores the information of an active selection under
+/// a [Selectable] or [SelectionHandler], including the start and
+/// end positions of a selection relative to the content, as well
+/// as an identifier to identify the owner of the content, any
+/// children selection ranges that may be contained within the active
+/// selection, and the content itself.
+///
+/// [SelectionArea] and [SelectableRegion] provide access to this
+/// information through the [SelectedContent] provided in their
+/// [SelectionArea.onSelectionChanged] and
+/// [SelectableRegion.onSelectionChanged] members.
+///
+/// {@tool dartpad}
+/// This example shows how to color red the active selection
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.1.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to replace the active selection
+/// under a [SelectionArea] or [SelectableRegion] with a widget.
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///   * [SelectedContent], which contains the [SelectedContentRange] for a given [Selectable] or [SelectionHandler].
+abstract class SelectedContentRange<T extends Object> {
+  /// Creates a range for the content of a [Selectable] or [SelectionHandler].
+  SelectedContentRange({
+    this.selectableId,
+    required this.content,
+    this.children,
+  });
+
+  /// The unique id for the [Selectable] that created the range.
+  final int? selectableId;
+
+  /// The content that contains the selection.
+  final T content;
+
+  /// The start of the selection relative to the [content].
+  int get startOffset;
+
+  /// The end of the selection relative to the [content].
+  int get endOffset;
+
+  /// Additional ranges to include as children.
+  ///
+  /// Children of a given range enable more granular modification of the
+  /// selection.
```

Comment:
Is it reasonable to expand on this here? After reading this I have questions like:

 * Are we talking about typically T being a TextSpan and the children being its children TextSpans?
 * Is there always a 1 to 1 mapping between SelectedContentRange and its `content`? E.g. one SelectedContentRange covers the entire range of one TextSpan, and all of its children TextSpans also each have a single SelectedContentRange that covers it entirely?
 * How do I know that those two trees are in sync and complete? Are they ever purposely partial? Or maybe this class is not concerned about that kind of thing and it's up to the owner?

Reply:
> Are we talking about typically T being a TextSpan and the children being its children TextSpans?

If T is a TextSpan then its children are any WidgetSpans that contain `Text` widgets. This doesn't include children `TextSpan`s because in the world of `SelectionArea` those are not children of the `Text` widget.

So for example say we had the following `Text` widget:

```dart
  const Text.rich(
    TextSpan(
      text: 'Some text in a textspan.',
      children: <InlineSpan>[
        TextSpan(text: 'some text in a child textspan.'),
        WidgetSpan(child: Text('some text in a widgetspan.')),
        TextSpan(text: 'some more text in a textspan.')
      ],
    ),
  );
```

Say we select everything in this `Text` widget, the `SelectedContentRange` that is given will include 1 child, which represents the text inside the `WidgetSpan`. The `content` member at the root `SelectedContentRange` will be the `TextSpan` that represents all the content in the `Text` widget including the `WidgetSpan`.

> Is there always a 1 to 1 mapping between SelectedContentRange and its content? E.g. one SelectedContentRange covers the entire range of one TextSpan, and all of its children TextSpans also each have a single SelectedContentRange that covers it entirely?

I would say no considering my answer to the previous question. You can say the `SelectedContentRange` follows the `Selectable` tree mapping and not the `TextSpan` tree mapping. If a given Selectable hasn't provided a `SelectedContentRange` for their given selection then they will not be included in this mapping.

> How do I know that those two trees are in sync and complete? Are they ever purposely partial? Or maybe this class is not concerned about that kind of thing and it's up to the owner?

If i'm understanding the question, they may be partial if a user has a custom selectable that does not provide a SelectedContentRange for its selected content. Within the confines of the framework they should always be complete since our only selectable at the moment is `Text`.

Reply:
I just wrote a previous comment saying that you should include an example in the top SelectedContentRange docs. I think this one you gave would also work besides the one I gave there, and I think it would clear up a lot of the questions I was asking here when users are trying to understand this class.

About my "in sync and complete" question. I think we have no guarantee here, but I think it's the same risk that we have that a Widget could modify its `children` List and make itself out of sync with say the RenderObject tree. It seems like Flutter would benefit from immutability here, but this PR is not the time to tackle that.